File: backend/app/schemas/deployment.py
"""
策略部署 Pydantic Schema

包含:
- 部署配置
- 参数范围限制
- 环境类型
"""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class DeploymentEnvironment(str, Enum):
    """部署环境"""
    PAPER = "paper"  # 模拟盘
    LIVE = "live"    # 实盘


class DeploymentStatus(str, Enum):
    """部署状态"""
    DRAFT = "draft"      # 草稿
    RUNNING = "running"  # 运行中
    PAUSED = "paused"    # 已暂停
    STOPPED = "stopped"  # 已停止


class StrategyType(str, Enum):
    """策略类型"""
    INTRADAY = "intraday"       # 日内交易
    SHORT_TERM = "short_term"   # 短线 (1-5天)
    MEDIUM_TERM = "medium_term" # 中线 (1-4周)
    LONG_TERM = "long_term"     # 长线 (>1月)


# ============ 参数范围定义 ============

class ParamRange(BaseModel):
    """参数范围"""
    min_value: float
    max_value: float
    default_value: float
    step: float = 0.01
    unit: str = ""
    description: str = ""


class RiskParams(BaseModel):
    """风控参数"""
    stop_loss: float = Field(-0.05, ge=-0.50, le=-0.01, description="止损比例")
    take_profit: float = Field(0.10, ge=0.02, le=1.0, description="止盈比例")
    max_position_pct: float = Field(0.10, ge=0.01, le=0.50, description="单只最大仓位")
    max_drawdown: float = Field(-0.15, ge=-0.50, le=-0.05, description="最大回撤限制")


class CapitalConfig(BaseModel):
    """资金配置"""
    total_capital: Decimal = Field(..., gt=0, description="总资金")
    initial_position_pct: float = Field(0.80, ge=0.10, le=1.0, description="初始仓位比例")
    reserve_cash_pct: float = Field(0.20, ge=0.0, le=0.50, description="预留现金比例")


# ============ 部署配置 ============

class DeploymentConfig(BaseModel):
    """部署配置"""
    # 基础信息
    strategy_id: str
    deployment_name: str = Field(..., min_length=1, max_length=100)
    environment: DeploymentEnvironment = DeploymentEnvironment.PAPER
    strategy_type: StrategyType = StrategyType.MEDIUM_TERM

    # 股票池配置 (继承自策略，可选择子集)
    universe_subset: Optional[list[str]] = None  # 为空则使用策略默认股票池

    # 风控参数 (继承自策略回测，可在范围内调整)
    risk_params: RiskParams = Field(default_factory=RiskParams)

    # 资金配置
    capital_config: CapitalConfig

    # 调仓设置
    rebalance_frequency: str = Field("daily", pattern="^(daily|weekly|monthly)$")
    rebalance_time: str = Field("09:35", pattern="^[0-2][0-9]:[0-5][0-9]$")


class DeploymentCreate(BaseModel):
    """创建部署请求"""
    config: DeploymentConfig
    auto_start: bool = False


class DeploymentUpdate(BaseModel):
    """更新部署请求"""
    deployment_name: Optional[str] = None
    risk_params: Optional[RiskParams] = None
    capital_config: Optional[CapitalConfig] = None
    rebalance_frequency: Optional[str] = None
    rebalance_time: Optional[str] = None


class Deployment(BaseModel):
    """部署实体"""
    deployment_id: str
    strategy_id: str
    strategy_name: str
    deployment_name: str
    environment: DeploymentEnvironment
    status: DeploymentStatus
    strategy_type: StrategyType

    # 配置
    config: DeploymentConfig

    # 运行时数据
    current_pnl: Decimal = Decimal("0")
    current_pnl_pct: float = 0
    total_trades: int = 0
    win_rate: float = 0

    # 时间戳
    created_at: datetime
    updated_at: datetime
    started_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class DeploymentListResponse(BaseModel):
    """部署列表响应"""
    total: int
    items: list[Deployment]


# ============ 参数范围限制 ============

class ParamLimits(BaseModel):
    """参数范围限制 (从策略回测结果继承)"""
    strategy_id: str

    # 风控参数范围
    stop_loss_range: ParamRange
    take_profit_range: ParamRange
    max_position_pct_range: ParamRange
    max_drawdown_range: ParamRange

    # 资金范围
    min_capital: Decimal = Field(Decimal("1000"), description="最低资金要求")

    # 股票池
    available_symbols: list[str] = Field(default_factory=list)

    class Config:
        from_attributes = True

File: frontend/src/types/deployment.ts
/**
 * 策略部署类型定义
 * Sprint 4 增强: 股票池配置、执行模式、前置检查
 */

// 部署环境
export type DeploymentEnvironment = 'paper' | 'live';

// 部署状态
export type DeploymentStatus = 'draft' | 'running' | 'paused' | 'stopped';

// 策略类型
export type StrategyType = 'intraday' | 'short_term' | 'medium_term' | 'long_term';

// ==================== Sprint 4 新增类型 ====================

// 执行模式 (F11)
export type ExecutionMode = 'auto' | 'confirm' | 'notify_only';

// 股票池配置模式
export type UniverseMode = 'full' | 'custom' | 'exclude';

// 股票池配置 (继承自策略，部署时不可修改)
export interface UniverseSubsetConfig {
  mode: UniverseMode;                   // 默认使用'full'继承策略配置
  maxPositions?: number;                 // 最大持股数量（继承自策略）
}

// 前置检查项 (F12)
export interface PreDeploymentCheck {
  id: string;
  name: string;
  description: string;
  status: 'pending' | 'passed' | 'failed' | 'warning';
  message?: string;
  required: boolean;  // 是否必须通过
}

// 前置检查结果 (F12)
export interface PreDeploymentCheckResult {
  allPassed: boolean;
  requiredPassed: boolean;
  checks: PreDeploymentCheck[];
}

// 执行模式配置 (F11)
export const EXECUTION_MODE_CONFIG: Record<ExecutionMode, {
  label: string;
  description: string;
  icon: string;
}> = {
  auto: {
    label: '全自动',
    description: '系统自动执行所有交易信号，无需人工确认',
    icon: '⚡'
  },
  confirm: {
    label: '确认执行',
    description: '每笔交易前需要您手动确认',
    icon: '👆'
  },
  notify_only: {
    label: '仅通知',
    description: '仅发送交易信号通知，不自动执行',
    icon: '🔔'
  },
};

// 股票池配置说明 (部署时继承策略配置)
// 注意: 股票池配置在策略构建器中定义，部署时自动继承，不再单独配置

// 参数范围
export interface ParamRange {
  minValue: number;
  maxValue: number;
  defaultValue: number;
  step: number;
  unit: string;
  description: string;
}

// 风控参数
export interface RiskParams {
  stopLoss: number;
  takeProfit: number;
  maxPositionPct: number;
  maxDrawdown: number;
}

// 资金配置
export interface CapitalConfig {
  totalCapital: number;
  initialPositionPct: number;
  reserveCashPct: number;
}

// 部署配置 (Sprint 4 扩展)
export interface DeploymentConfig {
  strategyId: string;
  deploymentName: string;
  environment: DeploymentEnvironment;
  strategyType: StrategyType;
  // F10: 股票池配置
  universeConfig: UniverseSubsetConfig;
  // F11: 执行模式
  executionMode: ExecutionMode;
  // 原有配置
  riskParams: RiskParams;
  capitalConfig: CapitalConfig;
  rebalanceFrequency: string;
  rebalanceTime: string;
}

// 默认股票池配置 (F10)
export const DEFAULT_UNIVERSE_CONFIG: UniverseSubsetConfig = {
  mode: 'full',
  maxPositions: 20,
};

// 默认执行模式 (F11)
export const DEFAULT_EXECUTION_MODE: ExecutionMode = 'confirm';

// 部署实体
export interface Deployment {
  deploymentId: string;
  strategyId: string;
  strategyName: string;
  deploymentName: string;
  environment: DeploymentEnvironment;
  status: DeploymentStatus;
  strategyType: StrategyType;
  config: DeploymentConfig;
  currentPnl: number;
  currentPnlPct: number;
  totalTrades: number;
  winRate: number;
  createdAt: string;
  updatedAt: string;
  startedAt?: string;
}

// 部署列表响应
export interface DeploymentListResponse {
  total: number;
  items: Deployment[];
}

// 参数范围限制
export interface ParamLimits {
  strategyId: string;
  stopLossRange: ParamRange;
  takeProfitRange: ParamRange;
  maxPositionPctRange: ParamRange;
  maxDrawdownRange: ParamRange;
  minCapital: number;
  availableSymbols: string[];
}

// 创建部署请求
export interface DeploymentCreateRequest {
  config: DeploymentConfig;
  autoStart?: boolean;
}

// 更新部署请求
export interface DeploymentUpdateRequest {
  deploymentName?: string;
  riskParams?: RiskParams;
  capitalConfig?: CapitalConfig;
  rebalanceFrequency?: string;
  rebalanceTime?: string;
}

// 状态配置
export const STATUS_CONFIG: Record<DeploymentStatus, {
  label: string;
  color: string;
  bgColor: string;
}> = {
  draft: { label: '草稿', color: 'gray', bgColor: 'bg-gray-500' },
  running: { label: '运行中', color: 'green', bgColor: 'bg-green-500' },
  paused: { label: '已暂停', color: 'orange', bgColor: 'bg-orange-500' },
  stopped: { label: '已停止', color: 'red', bgColor: 'bg-red-500' },
};

// 环境配置
export const ENV_CONFIG: Record<DeploymentEnvironment, {
  label: string;
  color: string;
  bgColor: string;
}> = {
  paper: { label: '模拟盘', color: 'blue', bgColor: 'bg-blue-500' },
  live: { label: '实盘', color: 'green', bgColor: 'bg-green-500' },
};

// 策略类型配置
export const STRATEGY_TYPE_CONFIG: Record<StrategyType, {
  label: string;
  holdingPeriod: string;
}> = {
  intraday: { label: '日内交易', holdingPeriod: '日内' },
  short_term: { label: '短线策略', holdingPeriod: '1-5天' },
  medium_term: { label: '中线策略', holdingPeriod: '1-4周' },
  long_term: { label: '长线策略', holdingPeriod: '>1月' },
};

// 默认风控参数
export const DEFAULT_RISK_PARAMS: RiskParams = {
  stopLoss: -0.05,
  takeProfit: 0.10,
  maxPositionPct: 0.10,
  maxDrawdown: -0.15,
};

// 默认资金配置
export const DEFAULT_CAPITAL_CONFIG: CapitalConfig = {
  totalCapital: 10000,
  initialPositionPct: 0.80,
  reserveCashPct: 0.20,
};
